File: Atividades/att3/Python/merge_sort.py
def merge(a, b):
    lena = len(a)
    lenb = len(b)
    index1, index2 = 0, 0

    res_list = list()

    while index1 < lena and index2 < lenb:
        if a[index1] < b[index2]:
            res_list.append(a[index1])
            index1 += 1
        
        else:
        # elif a[index1] >= b[index2]:
            res_list.append(b[index2])
            index2 += 1
    
    while index1 < lena:
        res_list.append(a[index1])
        index1 += 1

    while index2 < lenb:
        res_list.append(b[index2])
        index2 += 1
    
    return res_list

def half(n): return int(n / 2)

def merge_sort(lista):
    l = len(lista)
    ## uma implementação recursiva do merge_sort
    if l > 1:
        l2 = half(l)

        # particionando
        a1 = merge_sort(lista[0: l2])
        a2 = merge_sort(lista[l2: l])

        return merge(a1, a2)
    
    return lista

File: Atividades/att3/Go/mod/merge_sort.go
package sorters

func MergeSort(arr []int) []int {
	if len(arr) <= 1 {
		return arr
	}

	mid := len(arr) / 2
	left := MergeSort(arr[:mid])
	right := MergeSort(arr[mid:])

	return merge(left, right)
}

func merge(a, b []int) []int {
	lena := len(a)
	lenb := len(b)
	res_list := make([]int, 0, lena+lenb)
	i, j := 0, 0
	for i < lena || j < lenb {
		if i == lena {
			res_list = append(res_list, b[j])
			j++
		} else if j == lenb {
			res_list = append(res_list, a[i])
			i++
		} else if a[i] < b[j] {
			res_list = append(res_list, a[i])
			i++
		} else {
			res_list = append(res_list, b[j])
			j++
		}
	}
	return res_list
}
